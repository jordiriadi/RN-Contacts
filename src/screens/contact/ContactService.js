import Sql from '../../Sql/client';
import Contact from './types/Contact';

export function getContacts() {
  const extractPayload = (payload) => payload.data.map(Contact);

  return Sql.query({}, { method: 'GET' }).then(extractPayload);
}

export function addContact(formValues) {

  return Sql.query({}, { method: 'POST', query: formValues});
}

export function updateContact(id, _formValues) {
  const formValues = {
    firstName: _formValues.firstName,
    lastName: _formValues.lastName,
    age: _formValues.age
  };
  const stringifyParams = `/${id}`;

  return Sql.query({}, { method: 'PUT', query: formValues, stringifyParams})
}

export function deleteContact(id) {
  const stringifyParams = `/${id}`;

  return Sql.query({}, { method: 'DELETE', stringifyParams});
}

export default {
  getContacts,
  addContact,
  updateContact,
  deleteContact
}