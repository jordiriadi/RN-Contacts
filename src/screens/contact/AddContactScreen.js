import React from 'react';
import { StyleSheet, View } from 'react-native';

import ModifyContactForm from './ModifyContactForm';
import FlashMessage from '../../components/FlashMessage';
import ContactService from './ContactService';
import Styles from '../../utils/Styles';

class AddContactScreen extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
      error: null,
    }
  }

  handleAddContact = (formValues) => {
    const addContact = ContactService.addContact(formValues);

    const startLoading = () => this.setState({ loading: true });

    const handleSuccess = () => {
      this.props.navigation.navigate('ContactListScreen');
    }

    const handleFail = (err) => {
      this.setState({
        error: err,
        loading: false,
      });
      console.log(err);
    }

    startLoading();
    addContact.then(handleSuccess, handleFail).catch(handleFail);


    return addContact;
  }

  handleResetError = () => this.setState({ error: null });

  render() {
    const { error } = this.state;

    return (
      <View style={styles.container}>
        {error && (
          <FlashMessage
            text={error}
            onTimeout={this.handleResetError}
          />
        )}
        <ModifyContactForm
          modify={"create"}
          onSave={this.handleAddContact}
          isLoading={this.state.loading}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: Styles.DefaultPadding * 8,
    backgroundColor: Styles.PrimaryColor,
  }
})


export default AddContactScreen;