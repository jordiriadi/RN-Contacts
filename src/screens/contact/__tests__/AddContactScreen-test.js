import 'react-native';
import React from 'react';
import {shallow} from 'enzyme';

import AddContactScreen from '../AddContactScreen';
import ContactService from '../ContactService';

const mockData = {
  "age": "33",
  "firstName": "LukeTheSecond",
  "id": "b3abd640-b22g-11e8-b02f-cbfa15db428b",
  "lastName": "Skywalker", 
  "photo": "N/A"
}

ContactService.addContact = jest.fn(() => Promise.resolve(mockData));

describe('Add Contact Screen', () => {  
  it('renders correctly', () => {
  
    const wrapper = shallow(<AddContactScreen/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('can trigger handle add contact', () => {
    const component = shallow(<AddContactScreen/>);

    expect(component.instance().handleAddContact());
  });
})