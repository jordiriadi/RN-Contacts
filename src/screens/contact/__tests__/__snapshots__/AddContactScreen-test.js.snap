// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Contact Screen renders correctly 1`] = `ShallowWrapper {}`;
